Improved extension README (#2352)
posit-dev/publisher: extensions/vscode/README.md

# Posit Publisher - Deploy to Posit Connect

[Posit Publisher](https://github.com/posit-dev/publisher) is a code-first tool
for deploying to

> [Posit Connect](https://posit.co/products/enterprise/connect/).

Reviewer: Suggest including the types of things supported?

Posit Publisher is a code-first tool with a push-button user interface for deploying the things you build in R and Python to Posit Connect.

Supported content types (frameworks) (From the changelog?)
- Python content: APIs, applications, and notebooks
- Quarto content: .qmd and .Rmd files. Support for embedded Shiny apps
- R content: Shiny apps, RMarkdown, Plumber APIs, and Rmd with embedded Shiny apps

Author: I was a bit worried about listing everything since it would be quite long - see the types we support from the [Configuration doc](https://github.com/posit-dev/publisher/blob/main/docs/configuration.md#type).

But it did feel strange to leave it out. Thanks for the suggestion, I'll get something added!

Author: I included this verbatim. I tried to come up with a more succinct opener but its nice to include "code-first" and "push-button" into the first line. Let me know if you agree that it is a little cumbersome, and we can try to reduce it.

## Installation

Install Posit Publisher by using the Install button above, or using the
Extensions side bar in [Positron](https://github.com/posit-dev/positron) or

> [VS Code](https://code.visualstudio.com/) and searching for `posit.publisher`.

Author: Trying to be more generic about the install instructions here to account for users seeing this in Positron or VS Code.

Posit Publisher is available from the
[Visual Studio Code Marketplace](https://marketplace.visualstudio.com/items?itemName=Posit.publisher)
or from the [Open VSX Registry](https://open-vsx.org/extension/posit/publisher).

## Features

> ### Code-First Deployment

Reviewer: ```suggestion
### Code-first deployment
```
Style guide

Configure your project and then deploy your content to Posit Connect with a
click.

[Configuration files](https://github.com/posit-dev/publisher/blob/main/docs/configuration.md)
allow setting up how you want your content to look and run on Connect - from
choosing a title and description to
[Runtime process settings](https://docs.posit.co/connect/user/content-settings/#content-runtime).

Keep your content stable with language and tool versions set in your
configurations- all automatically detected based on your local environment.

Everything is saved under `.posit/` in your project directory for easy
re-deployment and collaboration.

> ### Seamless Collaboration

Reviewer: ```suggestion
### Seamless collaboration
```
style guide

Share your deployments and configurations with your team by committing the
`.posit/` directory to your version control system, allowing Collaborators to
easily pick up where you left off.

> ### Track all Deployments for a Project

Reviewer: ```suggestion
### Track all deployments for a project
```
style guide

> Separate staging and production deployments easily per project.

Reviewer: Clarify that this is an example?

Author: Add another line above, and clarified that this was an example.

> ### Deployment Logs

Reviewer: ```suggestion
### Deployment logs
```
style guide

Logs are organized to easily identify what went wrong.

<img
  src="https://camo.githubusercontent.com/a74aa7b8bc2d2a67508b51764757b4692dea2bb451a8f5f8e876d00e4125a09e/68747470733a2f2f63646e2e706f7369742e636f2f7075626c69736865722f6173736574732f696d672f6465706c6f796d656e742d6c6f67732d76696577322e706e67"
  alt="Deploying Logs with organized steps in the Posit Publisher Logs view"
/>

> ## Help and Feedback

Reviewer: ```suggestion
## Help and feedback
```
style guide

# Posit Publisher - Deploy to Posit Connect

[Posit Publisher](https://github.com/posit-dev/publisher) is a code-first tool

> with a push-button user interface for deploying the things you build in R and
> Python to [Posit Connect](https://posit.co/products/enterprise/connect/).

Reviewer 2: ```suggestion
with a push-button user interface for deploying the things you build in Python and
R to [Posit Connect](https://posit.co/products/enterprise/connect/).
```

Python first, but also 🔠

> Supported content includes:

Reviewer 2: ```suggestion
Supported frameworks or content types include:
```

- Python content: APIs, applications, and notebooks
- Quarto content: `.qmd` and `.Rmd` files. Support for embedded Shiny apps
- R content: Shiny apps, RMarkdown, Plumber APIs, and Rmd with embedded
  Shiny apps

## Installation

> Install Posit Publisher by using the **Install** button above, or using the
> **Extensions** side bar in [Positron](https://github.com/posit-dev/positron) or
> [VS Code](https://code.visualstudio.com/) and searching for `posit.publisher`.

Reviewer 2: ```suggestion
Install Posit Publisher using the **Install** button above, or using the
**Extensions** side bar in [Positron](https://github.com/posit-dev/positron) or
[VS Code](https://code.visualstudio.com/) and searching for `posit.publisher`.
```

Posit Publisher is available from the
[Visual Studio Code Marketplace](https://marketplace.visualstudio.com/items?itemName=Posit.publisher)
or from the [Open VSX Registry](https://open-vsx.org/extension/posit/publisher).

## Features

> ### Code-first deployment
> 
> Configure your content and then deploy to Posit Connect with a click.

Reviewer 2: ```suggestion
### Code-first deployment

Configure your content and then deploy to Posit Connect with ease.
```

Code-first and then "...with a click" is a bit dissonant

[Configuration files](https://github.com/posit-dev/publisher/blob/main/docs/configuration.md)

> allow setting up how you want your content to look and run on Connect - from

Reviewer 2: The UI here isn't great, but this is what looks like a hyphen to em-dash.

choosing a title and description to
[Runtime process settings](https://docs.posit.co/connect/user/content-settings/#content-runtime).

> Keep your content stable with language and tool versions set in your
> configurations — all automatically detected based on your local environment.

Reviewer 2: ```suggestion
Keep your content stable with language and tool versions set in your
configurations — all automatically detected based on your local environment.
And you also have the control to pin them to the exact version that works for you.
```

> Everything is saved in the `.posit/` directory for easy re-deployment and
> collaboration.

Reviewer 2: ```suggestion
Everything is saved in the `.posit/` directory for easy reproducibility and
collaboration.
```

### Seamless collaboration

Allow Collaborators to easily pick up where you left off by committing the
`.posit/` directory to your version control system. Your team can deploy using

> the same deployments and configurations.

Reviewer 2: IMHO, we might hold this for when we have a more dedicated story around this. I think the code / features are almost all there, but I'm not sure folks will read this and immediately get the process we are imagining here. It would be nice to have a how to or walkthrough describing that that we could link to.

Author: If I'm understanding the suggestion - this sounds like removing the "Seamless collaboration" feature section in this PR, and adding it in the future when we have those code / features and a document to walk users through the collaboration flow?

Reviewer 2: Yeah, that's what I was thinking. It's not a "you should absolutely do that" but I was thinking about it and what I would as a user take away from this and I'm not sure there's enough detail here (or elsewhere) that really shows off what this allows (and writing that how to will take a second, probably not before 4.2)

Author: I removed the section, and added a follow-up task to document our collaboration flow. As a part of that we can add it back to the feature list in this README

### Track all deployments for a project

Easily track all deployments for a project with Posit Publisher.

> For example, create separate deployments for staging and production.

Reviewer 2: Any more details we can add here? This seems a little sparse / only part of the story.

Also do we want "projects" here. In some ways it's a good use of that: if your'e deploying an API and then a front end that uses that, you can track both of those in the publisher in a way you couldn't really in RStudio before. But also if that's what we're calling `projects` that's not the staging/production thing...

Author: Yeah I struggled with this one. Separating out the multiple deployments and the project management features shouldn't be conflated like I was doing here.

Author: I split the two concepts into two sections and added a bit more detail
- "Track all content for a project"
  - Highlights the API / front end application use case
- "Multiple deployments and configurations"
  - Highlights the staging / production use case
  
Let me know if you think I captured those well

Reviewer 2: Looks good, thanks!

### Deployment logs

Logs are organized to easily identify what went wrong.

<img
  src="https://camo.githubusercontent.com/a74aa7b8bc2d2a67508b51764757b4692dea2bb451a8f5f8e876d00e4125a09e/68747470733a2f2f63646e2e706f7369742e636f2f7075626c69736865722f6173736574732f696d672f6465706c6f796d656e742d6c6f67732d76696577322e706e67"
  alt="Deploying Logs with organized steps in the Posit Publisher Logs view"

> />

Reviewer 2: I'm commenting this here mostly because I think these additions should go _above_ the logs section, but: 

What about adding sections describing: 

* Easily get from vscode/positron to the content on connect
* Adopt / associate a piece of content you've deployed form elsewhere (e.g. RStudio)

Author: I added both of the sections you suggest - great ideas. Let me know if you think I captured the details well underneath!

to deploy, and Posit Publisher will get you started with a configuration.

Once you are ready, deploy your content to Posit Connect.

### Code-first deployment

Configure your content with ease.

[Configuration files](https://github.com/posit-dev/publisher/blob/main/docs/configuration.md)
allow setting up how you want your content to look and run on Connect — from
choosing a title and description to
[Runtime process settings](https://docs.posit.co/connect/user/content-settings/#content-runtime).

Keep your content stable with language and tool versions set in your
configurations — all automatically detected based on your local environment.

Keep your content stable with language and tool versions set in your
configurations — all automatically detected based on your local environment.
And you also have the control to pin them to the exact version that works for
you.

Everything is saved in the `.posit/` directory for easy reproducibility and
collaboration.

### Track all content for a project

If your project requires multiple pieces of content, Posit Publisher can manage
them all inside VS Code or Positron. For example, deploy an API and an

> application to utilize it.

Reviewer 2: ```suggestion
application that utilizes it.
```

Minor: Feels a little bit more horse-cart than cart-horse to me
